crates/interpreter/src/interpreter/shared_stack.rs

```diff
@@ -0,0 +1,550 @@
+use crate::{
+    primitives::{B256, U256},
+    InstructionResult,
+};
+use alloc::vec::Vec;
+use core::fmt;
+
+/// EVM interpreter stack limit.
+pub(crate) const STACK_LIMIT: usize = 1024;
+
+/// EVM stack.
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct SharedStack {
+    /// Shared buffer
+    pub buffer: Vec<U256>,
+    /// Stack checkpoints for each depth
+    /// Invariant: these are always in bounds of `data`.
+    pub checkpoints: Vec<usize>,
+    /// Invariant: equals `self.checkpoints.last()`
+    pub last_checkpoint: usize,
+}
+
+impl fmt::Display for SharedStack {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("[")?;
+        for (i, x) in self.context_stack().iter().enumerate() {
+            if i > 0 {
+                f.write_str(", ")?;
+            }
+            write!(f, "{x}")?;
+        }
+        f.write_str("]")
+    }
+}
+
+impl Default for SharedStack {
+    #[inline]
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SharedStack {
+    /// Instantiate a new stack with the [default stack limit][STACK_LIMIT].
+    #[inline]
+    pub fn new() -> Self {
+        Self {
+            buffer: Vec::with_capacity(STACK_LIMIT),
+            checkpoints: Vec::with_capacity(32),
+            last_checkpoint: 0,
+        }
+    }
+
+    /// Prepares the shared stack for a new context
+    #[inline]
+    pub fn new_context(&mut self) {
+        let buf_len = self.buffer.len();
+        self.checkpoints.push(buf_len);
+        self.last_checkpoint = buf_len;
+        self.buffer.reserve(STACK_LIMIT);
```

Comment:
This reserve would copy the full buffer of the previous values, so it would be better to be more granular and introduce the Page of memory that would contain at least `4*1024` items. This value (as a 4) is very arbitrary but the idea is to reuse the current stack.

On `new_context` we would check if there are at least 1024 slots left on the present page, and if this is not the case would allocate a new page. `free_context` would do this in reverse.

Reply:
This seems exciting! Let me know if I got this right:
- instead of a single buffers we have multiple buffers (or "pages") of at least `4*1024` items -> which means a `Vec<Vec<U256>>` with initial value something like `vec![Vec::with_capacity(4 * STACK_LIMIT)]`
- on `new_context`, if there is enough space in the _last page_, we keep it for this new context, otherwise we allocate another page
- on `free_context`, we just mark that space of the _last page_ available in some way and then we make the necessary changes accordingly

In practice, this really resembles the model you proposed in https://github.com/ethereum/evmone/issues/481#issuecomment-1179088645. Since we know in advance how much we should expand the buffer (which was not possible with the shared memory)  this should play very nicely. It should have a little more overhead if you remain on the same context the whole tx since the setup is a little more complex than the previous one, but it should definitely be better for multi-contexts since we do not copy data anymore in order to get more capacity

Reply:
Exactly, this would be a bit more complex to do with `SharedMemory`, but for the stack, it is a lot nicer and we should do it now.